JavaScript/src/dateTime/baseHoliday.ts

```diff
@@ -0,0 +1,247 @@
+                ret.futureValue = ret.pastValue = new Date(year, value.getMonth(), value.getDate());
+                ret.success = true;
+                return ret;
+            }
+
+            ret.timex = "XXXX" + timexStr;
+            ret.futureValue = this.getFutureValue(value, referenceDate, holidayKey);
+            ret.pastValue = this.getPastValue(value, referenceDate, holidayKey);
+            ret.success = true;
+
+            return ret;
+        }
+
+        return ret;
+    }
+
+    private getFutureValue(value: Date, referenceDate: Date, holiday: string): Date {
+        if (value < referenceDate) {
+            let func = this.config.holidayFuncDictionary.get(holiday);
+            if (func) {
+                return func(value.getFullYear() + 1);
+            }
+        }
+        return value;
+    }
+
+    private getPastValue(value: Date, referenceDate: Date, holiday: string): Date {
+        if (value >= referenceDate) {
+            let func = this.config.holidayFuncDictionary.get(holiday);
+            if (func) {
+                return func(value.getFullYear() - 1);
+            }
+        }
+        return value;
+    }
+}
+
+export abstract class BaseHolidayParserConfiguration implements IHolidayParserConfiguration {
+    variableHolidaysTimexDictionary: ReadonlyMap<string, string>;
+    holidayFuncDictionary: ReadonlyMap<string, (year: number) => Date>;
+    holidayNames: ReadonlyMap<string, string[]>;
+    holidayRegexList: RegExp[];
+    abstract getSwiftYear(text: string): number;
+    abstract sanitizeHolidayToken(holiday: string): string;
+
+    constructor() {
+        this.variableHolidaysTimexDictionary = BaseDateTime.VariableHolidaysTimexDictionary;
+        this.holidayFuncDictionary = this.initHolidayFuncs();
+    }
+
+    protected initHolidayFuncs(): ReadonlyMap<string, (year: number) => Date> {
+        return new Map<string, (year: number) => Date>(
+            [
+                ["fathers", BaseHolidayParserConfiguration.FathersDay],
+                ["mothers", BaseHolidayParserConfiguration.MothersDay],
+                ["thanksgivingday", BaseHolidayParserConfiguration.ThanksgivingDay],
+                ["thanksgiving", BaseHolidayParserConfiguration.ThanksgivingDay],
+                ["martinlutherking", BaseHolidayParserConfiguration.MartinLutherKingDay],
+                ["washingtonsbirthday", BaseHolidayParserConfiguration.WashingtonsBirthday],
+                ["canberra", BaseHolidayParserConfiguration.CanberraDay],
+                ["labour", BaseHolidayParserConfiguration.LabourDay],
+                ["columbus", BaseHolidayParserConfiguration.ColumbusDay],
+                ["memorial", BaseHolidayParserConfiguration.MemorialDay]
+            ]);
+    }
+
+    // All months are zero-based (-1)
+    private static MothersDay(year: number): Date { return new Date(year, 5 - 1, BaseHolidayParserConfiguration.getDay(year, 5 - 1, 1, DayOfWeek.Sunday)); }
```

Comment:
We should think of a way to make this auto-generated in the future.
Can you add a TODO here?